### PR TITLE
Feature: Sandboxes Monitoring - 2 - Testing Infrastructure

### PR DESCRIPTION
**Pull request by ben-fornefeld**
- [ ] 1: API/State Layer #137 
- [ ] 2: **THIS PR**

Builds upon the API/State Management layer to provide comprehensive test coverage for all utility functions and data processing logic. Integrates unit tests into the CI pipeline for early regression detection.

## Key Features

**Comprehensive Unit Test Coverage (104 tests)**
- **Chart utilities** (25 tests) - Data transformation, axis calculations, tooltip formatting
- **Formatting functions** (47 tests) - Date/time/number formatting across locales and formats
- **Sandboxes utilities** (17 tests) - Metrics processing and transformation logic
- **Team metrics calculations** (15 tests) - Step calculations and aggregation algorithms

**Development & Load Testing**
- `burst.test.ts` - Tests system behavior under sudden load spikes
- `traffic.test.ts` - Tests sustained traffic patterns and performance
- Configurable test scenarios for different usage patterns

**GitHub Actions CI Integration**
- Dedicated `unit-tests` job that runs before integration tests
- Minimal environment setup (no external services required)
- Fast feedback loop (~1 second execution time)
- Cost-efficient CI pipeline (integration tests only run if unit tests pass)

**Test Infrastructure**
- Supporting UI components needed for testing (chart-utils, tooltips)
- Proper test isolation and mocking strategies
- Enhanced test documentation and examples

### CI Pipeline Flow

```mermaid
graph LR
    subgraph "GitHub Actions"
        Push[Push/PR] --> UnitTests["Unit Tests<br/>104 tests<br/>~1s execution"]
        UnitTests --> |"✅ Pass"| IntegrationTests["Integration Tests<br/>~30s execution"]
        UnitTests --> |"❌ Fail"| Stop[❌ Stop Pipeline]
        IntegrationTests --> |"✅ Pass"| Success[✅ All Tests Pass]
        IntegrationTests --> |"❌ Fail"| Fail[❌ Integration Failure]
    end

    subgraph "Test Categories"
        UnitTests --> ChartUtils["Chart Utils (25)"]
        UnitTests --> Formatting["Formatting (47)"]
        UnitTests --> Sandboxes["Sandboxes (17)"]
        UnitTests --> Metrics["Team Metrics (15)"]
    end

    style UnitTests fill:#e1f5fe
    style IntegrationTests fill:#e8f5e9
    style Success fill:#c8e6c9
    style Fail fill:#ffcdd2
    style Stop fill:#ffcdd2
```

### Test Coverage Map

```mermaid
graph TB
    subgraph "Utility Functions Tested"
        subgraph "Chart Utilities"
            CalcAvg["calculateAverage()"]
            CalcMedian["calculateMedian()"]
            CalcCentralTendency["calculateCentralTendency()"]
            CalcYAxis["calculateYAxisMax()"]
            FillZeros["fillMetricsWithZeros()"]
            TransformData["transformMetricsToLineData()"]
            CreateSeries["createChartSeries()"]
        end
        
        subgraph "Formatting Functions"
            FormatNumber["formatNumber()"]
            FormatBytes["formatBytes()"]
            FormatDuration["formatDuration()"]
            FormatDate["formatChartTimestamp*()"]
            FormatCompact["formatCompactNumber()"]
            FormatDecimal["formatDecimal()"]
            FormatPercentage["formatPercentage()"]
        end
        
        subgraph "Data Processing"
            TransformMetrics["transformMetricsToClientMetrics()"]
            CalcStep["calculateStepForDuration()"]
            FillTeamMetrics["fillTeamMetricsWithZeros()"]
        end
    end

    subgraph "Test Files"
        ChartTests["chart-utils.test.ts<br/>25 tests"]
        FormatTests["formatting.test.ts<br/>47 tests"] 
        SandboxTests["sandboxes.test.ts<br/>17 tests"]
        MetricsTests["team-metrics-step.test.ts<br/>15 tests"]
    end

    ChartTests --> CalcAvg
    ChartTests --> CalcMedian
    ChartTests --> CalcCentralTendency
    ChartTests --> CalcYAxis
    ChartTests --> FillZeros
    ChartTests --> TransformData
    ChartTests --> CreateSeries

    FormatTests --> FormatNumber
    FormatTests --> FormatBytes
    FormatTests --> FormatDuration
    FormatTests --> FormatDate
    FormatTests --> FormatCompact
    FormatTests --> FormatDecimal
    FormatTests --> FormatPercentage

    SandboxTests --> TransformMetrics
    SandboxTests --> CalcStep
    
    MetricsTests --> FillTeamMetrics

    style ChartTests fill:#e1f5fe
    style FormatTests fill:#fff3e0
    style SandboxTests fill:#e8f5e9
    style MetricsTests fill:#f3e5f5
```

## Technical Implementation

**Test Architecture**
- Vitest as the test runner with TypeScript support
- Tests focus on pure functions and data transformations
- Mock data generation for consistent test scenarios
- Comprehensive edge case coverage (empty data, boundary conditions, timezone handling)

**Key Test Files**
- `src/__test__/unit/chart-utils.test.ts` - Chart calculation logic
- `src/__test__/unit/formatting.test.ts` - All formatting utilities
- `src/__test__/unit/sandboxes.test.ts` - Data transformation functions
- `src/__test__/unit/team-metrics-step.test.ts` - Time interval calculations
- `src/__test__/development/` - Load and performance testing

**CI Pipeline Enhancements**
- Unit tests run first for fast feedback
- Integration tests depend on unit test success
- Parallel test execution for optimal performance
- Clear separation of test types and responsibilities

---
*This PR provides comprehensive test coverage for the API/State Management layer. All 104 unit tests pass and are integrated into CI. The testing infrastructure validates that all utility functions work correctly and provides confidence for further development. Requires PR1 (API layer) to be merged first.*

### CODE REVIEW

File: src/features/dashboard/sandboxes/monitoring/chart-utils.tsx [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+/**
+ * Fill gaps in team metrics with zeros for charting
+ * Detects anomalous gaps and adds zeros at start/end and between data waves
+ * The step parameter is the expected/display step - we detect when actual data
+ * intervals are larger than this and fill the gaps
+ *
+ * Note: Backend overfetches by one step to capture boundary points,
+ * so if the last point is close to the end, it likely means activity continued
+ */
+export function fillMetricsWithZeros(
```
Comment (mishushakov) [inline src/features/dashboard/sandboxes/monitoring/chart-utils.tsx:225]: please break this function down into multiple, it feels way to big

File: src/ui/data/chart-utils.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  return (params: echarts.TooltipComponentFormatterCallbackParams) => {
+    const paramsData = Array.isArray(params) ? params[0] : params
+    if (!paramsData?.value) return ''
+
+    const value = Array.isArray(paramsData.value)
+      ? paramsData.value[1]
+      : paramsData.value
+    const timestamp = Array.isArray(paramsData.value)
+      ? (paramsData.value[0] as string)
+      : (paramsData.value as string)
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply src/ui/data/chart-utils.tsx:39]: maybe split into two functions, one for line charts we use (array) and for the other one and then another function which will on the input shape decide what to return

File: src/ui/data/line-chart.utils.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import deepmerge from 'deepmerge'
+import { EChartsOption } from 'echarts'
+import { defaultLineChartOption } from './line-chart.defaults'
+
+export const mergeReplaceArrays = <T>(target: T, ...sources: Partial<T>[]): T =>
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  deepmerge.all([target as any, ...sources] as any, {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    arrayMerge: (_destinationArray: any[], sourceArray: any[]) => sourceArray,
+  }) as T
```
Comment (mishushakov) [inline src/ui/data/line-chart.utils.ts:10]: I think in TS there's Array type, maybe you can use this instead of any[]

Comment (mishushakov) [reply src/ui/data/line-chart.utils.ts:10]: btw. if you're only using this in buildLineChartOption, then maybe move it there & use echarts types

File: src/features/dashboard/sandboxes/monitoring/chart-utils.tsx (hunk 1 above)
Comment (mishushakov) [reply src/features/dashboard/sandboxes/monitoring/chart-utils.tsx:225]: also, make sure there are no side-effects when manipulating the arrays

**Review by mishushakov — CHANGES_REQUESTED** [review]
minor nits, otherwise good to go
